### PR TITLE
Add the ability to unseal using recovery keys via an explicit seal option.

### PR DESCRIPTION
**Pull request by sgmiller**
At initialization time, we use the combined recovery key to store a second
copy of the barrier key at a second location if an auto seal is configured.
If the auto seal is in a failed state, the operator can stop vault, add the
`recovery_mode` bool to the seal stanza and restart.  They then unseal
Vault with recovery keys.

### CODE REVIEW

File: command/server/config.go [hunk 1]
```diff
@@ -464,13 +464,18 @@ func CheckConfig(c *Config, e error) (*Config, error) {
...
+	switch len(c.Seals) {
+	case 2:
 		switch {
 		case c.Seals[0].Disabled && c.Seals[1].Disabled:
 			return nil, errors.New("seals: two seals provided but both are disabled")
 		case !c.Seals[0].Disabled && !c.Seals[1].Disabled:
 			return nil, errors.New("seals: two seals provided but neither is disabled")
 		}
+	case 1:
+		if c.Seals[0].Disabled && c.Seals[0].Recover {
```
Comment (cipherboy) [inline command/server/config.go:476]: Might be worth adding a note why this isn't an issue for the two-seal case (I believe, since the recovery seal isn't necessary for operation, it could be disabled but the other (perhaps auto-unseal) one is enabled).

File: changelog/18683.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+```release-note:improvement
```
Comment (ncabatoff) [inline changelog/18683.txt:1]:
```suggestion
```release-note:change
```

It is an improvement, but assuming we go with opt-out approach, we should use the loudest changelog option.

File: command/server/config.go (hunk 1 above)
Comment (ncabatoff) [reply command/server/config.go:476]: Related: do we intend for recovery seal to be used in conjunction with seal migration?

Comment (ncabatoff) [reply command/server/config.go:476]: I think the answer is no, but it would be nice if that could be supported, e.g., I've lost my aws key, please help me migrate away from it using the recovery keys.

File: internalshared/configutil/kms.go [hunk 3]
```diff
@@ -99,6 +100,15 @@ func parseKMS(result *[]*KMS, list *ast.ObjectList, blockName string, maxKMS int
 			delete(m, "disabled")
 		}
 
+		var recover bool
+		if v, ok := m["recovery_mode"]; ok {
```
Comment (ncabatoff) [inline internalshared/configutil/kms.go:104]: I worry a tiny bit about confusion with the existing "recovery mode" in vault: https://developer.hashicorp.com/vault/docs/concepts/recovery-mode.  Suggested alternative: enable_recovery_keys?

File: vault/core.go [hunk 4]
```diff
@@ -2687,6 +2697,10 @@ func (c *Core) adjustForSealMigration(unwrapSeal Seal) error {
 			// We have the same barrier type and the unwrap seal is nil so we're not
 			// migrating from same to same, IOW we assume it's not a migration.
 			return nil
+		case c.seal.BarrierType() == wrapping.WrapperTypeShamir && isUnsealRecoverySeal(c.seal):
```
Comment (ncabatoff) [inline vault/core.go:2700]: When does it make sense to migrate to a recovery seal?

File: vault/init.go [hunk 5]
```diff
@@ -386,6 +386,21 @@ func (c *Core) Initialize(ctx context.Context, initParams *InitParams) (*InitRes
 			}
 
 			results.RecoveryShares = recoveryUnsealKeys
+
+			keysToStore := [][]byte{barrierKey}
+			if err := c.seal.SetStoredKeys(ctx, keysToStore); err != nil {
```
Comment (ncabatoff) [inline vault/init.go:391]: Why is this needed?

File: internalshared/configutil/kms.go (hunk 3 above)
Comment (sgmiller) [reply internalshared/configutil/kms.go:104]: Yeah, I was struggling with verbiage too.  I was also considering `recovery_keys_fallback`.

File: vault/seal.go [hunk 6]
```diff
@@ -40,6 +42,9 @@ const (
 	// recoveryKeyPath is the path to the recovery key
 	recoveryKeyPath = "core/recovery-key"
 
+	// recoveryUnsealKeyPath is the path to the recovery unseal key
```
Comment (ncabatoff) [inline vault/seal.go:45]:
```suggestion
	// recoveryUnsealKeyPath is the path to a copy of the root key,
	// encrypted using the shamir-combined recovery keys, just like
	// StoredBarrierKeysPath is the root key encrypted by the seal
	// (which in the case of a shamir seal is the shamir-combined unseal keys.)
```

File: command/server/config.go (hunk 1 above)
Comment (ncabatoff) [reply command/server/config.go:476]: Ok, on closer review it looks like the answer is yes, but then I'm not sure this is right.  I would say we shouldn't allow for a single seal to be disabled, and that in a migration (two seal) scenario, it doesn't make sense to have the non-disabled seal (the migration target) have recovery keys enabled.

**Review by ncabatoff — COMMENTED** [review]
Tests and docs are still forthcoming it seems?

File: internalshared/configutil/kms.go (hunk 3 above)
Comment (ncabatoff) [reply internalshared/configutil/kms.go:104]: That seems better, less ambiguous than my suggestion.

File: vault/core.go (hunk 4 above)
Comment (sgmiller) [reply vault/core.go:2700]: It doesn't, but adjustForSealMigration is called regardless of whether a migration is occurring or not, and without allowing this case, it rejects the seal in recovery mode and errors out.

File: vault/init.go (hunk 5 above)
Comment (sgmiller) [reply vault/init.go:391]: Oops, it isn't. That was from one of many previous attempts.

File: vault/core.go (hunk 4 above)
Comment (ncabatoff) [reply vault/core.go:2700]: Mind adding a comment?

File: command/server/config.go (hunk 1 above)
Comment (sgmiller) [reply command/server/config.go:476]: I was not envisioning allowing migration while doing recovery unseal, but you make a good point.  I should at least guard against recovery being enabled on the target seal.

**Comment by sgmiller** [comment]
> Tests and docs are still forthcoming it seems?

Yes.  Separate tickets for them.  Wanted to make sure the approach was sound as the tests will be a little tricky.

File: vault/core.go (hunk 4 above)
Comment (sgmiller) [reply vault/core.go:2700]: sure

File: changelog/18683.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+```release-note:change
+core: recovery keys may now be used to unseal an auto-sealed Vault via an
+opt in unseal recovery mode.
```
Comment (ncabatoff) [inline changelog/18683.txt:3]:
```suggestion
opt-out unseal recovery mode.
```

File: vault/seal.go [hunk 8]
```diff
@@ -110,6 +134,10 @@ func (d *defaultSeal) SetAccess(access *seal.Access) {
 
 func (d *defaultSeal) SetCore(core *Core) {
 	d.core = core
+	if d.logger == nil {
+		d.logger = d.core.Logger().Named("defaultseal")
```
Comment (ncabatoff) [inline vault/seal.go:138]: We could give this a better name (e.g. shamir-seal or autoseal-recovery) by looking at unsealKeyPath, right?

File: vault/seal.go [hunk 9]
```diff
@@ -197,8 +225,10 @@ func (d *defaultSeal) BarrierConfig(ctx context.Context) (*SealConfig, error) {
 		conf.Type = d.BarrierType().String()
 	case d.BarrierType().String():
 	default:
-		d.core.logger.Error("barrier seal type does not match expected type", "barrier_seal_type", conf.Type, "loaded_seal_type", d.BarrierType())
-		return nil, fmt.Errorf("barrier seal type of %q does not match expected type of %q", conf.Type, d.BarrierType())
+		if !(conf.Type != wrapping.WrapperTypeShamir.String() && d.unsealKeyPath == recoveryUnsealKeyPath) {
```
Comment (ncabatoff) [inline vault/seal.go:228]: You could eliminate the `if` by replacing `default` with the condition.  Also, would it be clearer if you expanded the negation and eliminated the parens, i.e. `conf.Type == wrapping.WrapperTypeShamir.String(), d.unsealKeyPath != recoveryUnsealKeyPath`?

File: vault/seal.go [hunk 10]
```diff
@@ -267,7 +297,7 @@ func (d *defaultSeal) RecoveryConfig(ctx context.Context) (*SealConfig, error) {
 }
 
 func (d *defaultSeal) RecoveryKey(ctx context.Context) ([]byte, error) {
-	return nil, fmt.Errorf("recovery not supported")
+	return d.access.Wrapper.(*aead.ShamirWrapper).KeyBytes(ctx)
```
Comment (ncabatoff) [inline vault/seal.go:300]: It feels like this behaviour change is going to be confusing, esp in conjuction with VerifyRecoveryUnsealKey and SetRecoveryKey as defined here.  Maybe we should introduce a new method for this instead?

File: vault/seal.go (hunk 9 above)
Comment (sgmiller) [reply vault/seal.go:228]: The switch is over conf.Type, so not quite that simple.  I'll invert the predicate though, simple enough.

**Comment by maxb** [comment]
It's wonderful to see this is being worked on! You can mark it as closing the issue #15490 I opened, when merged.

A point of curiosity: How is this going to work in Enterprise, given this change seems like it can't be compatible with Seal Wrap? I guess administrators will have to choose which feature they opt for?

**Comment by sgmiller** [comment]
> It's wonderful to see this is being worked on! You can mark it as closing the issue #15490 I opened, when merged.
> 
> A point of curiosity: How is this going to work in Enterprise, given this change seems like it can't be compatible with Seal Wrap? I guess administrators will have to choose which feature they opt for?

We have plans for tackling ENT use cases down the road.  This is the simpler first step.

File: vault/external_tests/sealmigration/testshared.go [hunk 11]
```diff
@@ -120,6 +124,58 @@ func ParamTestSealMigrationTransitToShamir_Post14(t *testing.T, logger hclog.Log
...
+	}()
+	sealKeyName := "transit-seal-key-1"
+	tss.MakeKey(t, sealKeyName)
+
+	// Initialize the backend with transit.
+	cluster, opts := InitializeTransit(t, logger, storage, basePort, tss, sealKeyName)
+	rootToken, recoveryKeys := cluster.RootToken, cluster.RecoveryKeys
+
+	// Disable the transit seal, forcing recovery
+	tss.Cleanup()
```
Comment (ncabatoff) [inline vault/external_tests/sealmigration/testshared.go:143]: Feels like we should be doing `tss = nil` here, and removing the tss.Cleanup below.

Comment (sgmiller) [reply vault/external_tests/sealmigration/testshared.go:143]: We need to shut down Transit so that it's unreachable and can't unseal without the recovery key method.  nilling it leaves the transit vault up with its port open right?

**Review by ncabatoff — APPROVED** [review]

File: vault/external_tests/sealmigration/testshared.go (hunk 11 above)
Comment (ncabatoff) [reply vault/external_tests/sealmigration/testshared.go:143]: I mean after calling Cleanup, assign nil.

Comment (sgmiller) [reply vault/external_tests/sealmigration/testshared.go:143]: ahh

**Comment by prudnitskiy** [comment]
Are there any estimations on including this in the release version?

Will this feature available in Vault OSS?

**Comment by solutiongeek** [comment]
During the development of this feature and corresponding enterprise enhancements, we determined that adoption of this version of the solution would create a potential conflict when migrating to enterprise or vice versa. For example, the original implementation was incompatible with seal wrap, so we withdrew this for a more holistic solution that does not have this inconsistency, which could lead to data loss in some scenarios if the seal wrapped values could no longer be decrypted. Currently we do not have any plans to bring this feature into OSS. We welcome any feedback or suggestions related to OSS seal recovery options due to non-availability of KMS.

**Comment by jdloft** [comment]
> Currently we do not have any plans to bring this feature into OSS.

Which feature? Seal wrap or the "more holistic" replacement for this?

**Comment by maxb** [comment]
> the original implementation was incompatible with seal wrap

I did wonder how people were planning to deal with that - I even called it out in an earlier comment (https://github.com/hashicorp/vault/pull/18683#issuecomment-1401284058) before this was merged :-)

@solutiongeek, are you able to share a basic outline of what this "more holistic solution" is?

Right now, OSS users have unfortunately been tempted with a tantalizing prospect of this long standing issue being sorted, only for it to be snatched away. It would be nice to figure out a way forward.

**Comment by xskrasek** [comment]
Having the option to use recovery keys for unsealing is really useful. I would also like to know the current plans after the recent solution was discarded.

**Comment by solutiongeek** [comment]
Right now, I don't have better answers to your questions that I can share, beyond what I've said above. The team will update when there is more to share. If you have strong opinions or would like to share your experiences with the issue, please reach out and my team will be happy to talk with you and share your feedback. Thank you.

**Comment by jrutkowski** [comment]
> We welcome any feedback or suggestions related to OSS seal recovery options due to non-availability of KMS.

Let's imagine the following scenario:
- When using auto-unseal configuration option, Vault still generates an unseal key, based on parameters provided by the operator (number of shares and the threshold).
- This key is encrypted by the cloud provider backend (KMS etc.) and stored in the storage, as it is right now, but with the redundancy resulting from Shamir's Secret Sharing algorithm used.
- If the operator directly requests it (during the initialization), the same set of keys is returned to the operator, so they can be later used for unsealing the Vault if the access to cloud provider's cryptographic backend is for some reason lost.

Is there anything wrong with that approach? I'm not an cryptographic expert, but this seems to me like an easy and straight-forward solution.
Even if it would introduce some risks - would they be bigger, than the risk or having all the data and backups completely inaccessible when loosing the access to cloud provider's cryptographic service?

**Comment by maxb** [comment]
Really what we need is just to have the solution that was previously implemented and reverted, put back. It was a good design, it was just fundamentally incompatible with Seal Wrap (because the very definition of Seal Wrap is that certain parts of the datastore are encrypted directly via the HSM). A good way forward would be to just accept and document that, along with code safeguards to ensure incompatible configurations are rejected.

**Comment by xskrasek** [comment]
Having this feature would be incredibly helpful. In case this is not a viable solution, the recovery keys should be, at the very least renamed, as the name is incredibly misleading.

**Comment by Dragubitt** [comment]
I'm agree with @maxb this feature would be very helpful. 

Then it's almost mandatory to use Autounseal based on Cloud provider KMS (AWS KMS, ...), cause it is not possible to restore backup or even unseal Vault when key is unavailable (deleted accidentaly or not other disaster)

**Comment by NagenderPulluri** [comment]
I have vault deployed in region1 and for auto unseal I used AWS KMS key(single region specific), as part of DR i'm restoring it in region2, but couldn't able to unseal the vault as the KMS key used not exists in region2, how to resolve this ?

**Comment by jrutkowski** [comment]
> I have vault deployed in region1 and for auto unseal I used AWS KMS key(single region specific), as part of DR i'm restoring it in region2, but couldn't able to unseal the vault as the KMS key used not exists in region2, how to resolve this ?

Technically, you can spawn another instance of vault in region1 and use it to "reseal" the underlying database from KMS key, to the set of shamir keys. Then make a snapshot, import it in region2 and "reseal" again with a different KMS key. Far from the ideal solution though...

**Comment by sleeptillseven** [comment]
From what I understand, this feature will go directly  into Enterprise (and not OSS), is that correct. Any updates?

**Comment by 131** [comment]
@maxb , @sgmiller , as this issue is mostly impeding vault enterprise, could you rebase this into openbao ? (as its main purpose is to remove the enterprise handbrake) ?
